internal/future/app/mmlogic/mmlogic_service.go

```diff
@@ -43,21 +55,51 @@ func newMmlogic(cfg config.View) (*mmlogicService, error) {
 	return ms, nil
 }
 
-// GetPoolTickets gets the list of Tickets that match every Filter in the
+// QueryTickets gets the list of Tickets that match every Filter in the
 // specified Pool.
-func (s *mmlogicService) QueryTickets(req *pb.QueryTicketsRequest, stream pb.MmLogic_QueryTicketsServer) error {
-	ctx := stream.Context()
+func (s *mmlogicService) QueryTickets(req *pb.QueryTicketsRequest, responseServer pb.MmLogic_QueryTicketsServer) error {
+	poolFilters := req.Pool.Filter
+
+	// Send requests to the storage service
+	idsToProperties, err := s.store.FilterTickets(poolFilters)
+	if err != nil {
+		logger.WithError(err).Error("Failed to retrieve result from storage service.")
+		return status.Error(codes.Internal, err.Error())
+	}
 
-	for {
+	ctx := responseServer.Context()
+	idx := 0
+	// The number of tickets in a paging response
+	pSize := s.cfg.GetInt("storage.page.size")
+	page := []*pb.Ticket{}
+
+	for id, property := range idsToProperties {
 		select {
 		case <-ctx.Done():
 			return nil
-
 		default:
-			err := stream.Send(&pb.QueryTicketsResponse{})
+			propertyByte, err := json.Marshal(property)
 			if err != nil {
-				return err
+				logger.WithError(err).Error("Failed to convert go struct to JSON")
+				return status.Errorf(codes.Internal, err.Error())
+			}
+			propertyJSON := string(propertyByte)
+			page = append(page, &pb.Ticket{Id: id, Properties: propertyJSON})
+
+			idx++
+
+			endPage := idx%pSize == 0 || idx == len(idsToProperties)-1
+			if endPage {
+				// Reaches page limit; Send a stream response then reset the page
+				err := responseServer.Send(&pb.QueryTicketsResponse{Ticket: page})
+				if err != nil {
+					logger.WithError(err).Error("Failed to send Redis response to grpc server")
+					return status.Errorf(codes.Unknown, err.Error())
```

Comment:
Internal ? Or Unknown?

Reply:
I'll probably do Abort. It make sense to return a grpc abort error if something goes wrong when sending/receiving messages.

Reply:
From the description here, looks like Aborted is if the client must retry - https://github.com/grpc/grpc-go/blob/master/codes/codes.go - Do you think retrying will help here? Internal is when something unexpected happens - indicating some issue with the system itself. @jeremyje might have guidance here.